PolyEngine/RenderingDevice/OpenGL/Src/Pipeline/EnvCapture.cpp

```diff
@@ -0,0 +1,311 @@
+{
+	// create cube map resource for capture
+	gConsole.LogInfo("EnvCapture::CaptureCubemap create cubemap resource");
+	glGenTextures(1, &EnvCubemap);
+	glBindTexture(GL_TEXTURE_CUBE_MAP, EnvCubemap);
+	for (GLuint i = 0; i < 6; ++i)
+	{
+		// note that we store each face with 16 bit floating point values
+		glTexImage2D(GL_TEXTURE_CUBE_MAP_POSITIVE_X + i, 0, GL_RGB16F, 512, 512, 0, GL_RGB, GL_FLOAT, nullptr);
+	}
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
+	// glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	gConsole.LogInfo("EnvCapture::CaptureCubemap cubemap resource created");
+	glBindTexture(GL_TEXTURE_CUBE_MAP, 0);
+	CHECK_GL_ERR();
+
+	// Create FBO for HDR capture
+	gConsole.LogInfo("EnvCapture::CaptureCubemap Create capture FBO");
+	GLuint captureFBO;
+	GLuint captureRBO;
+	glGenFramebuffers(1, &captureFBO);
+	glGenRenderbuffers(1, &captureRBO);
+
+	glBindFramebuffer(GL_FRAMEBUFFER, captureFBO);
+	glBindRenderbuffer(GL_RENDERBUFFER, captureRBO);
+	glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, 512, 512);
+	glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, captureRBO);
+	gConsole.LogInfo("EnvCapture::CaptureCubemap Capture FBO created");
+	CHECK_GL_ERR();
+
+	// Projection
+	gConsole.LogInfo("EnvCapture::CaptureCubemap create matrices");
+	Matrix uClipFromView;
+	uClipFromView.SetPerspective(90.0_deg, 1.0f, 0.1f, 10.0f);
+
+	Matrix ViewFromModel[] =
+	{
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(1.0f,  0.0f,  0.0f), Vector(0.0f, -1.0f,  0.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(-1.0f,  0.0f,  0.0f), Vector(0.0f, -1.0f,  0.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f, -1.0f,  0.0f), Vector(0.0f,  0.0f, -1.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f,  1.0f,  0.0f), Vector(0.0f,  0.0f,  1.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f,  0.0f,  1.0f), Vector(0.0f, -1.0f,  0.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f,  0.0f, -1.0f), Vector(0.0f, -1.0f,  0.0f))
+	};
+	gConsole.LogInfo("EnvCapture::CaptureCubemap matrices created");
+
+	// convert HDR equirectangular environment map to cubemap equivalent
+	gConsole.LogInfo("EnvCapture::CaptureCubemap start cubemap rendering loop");
+	EquirectangularToCubemapShader.BindProgram();
+	EquirectangularToCubemapShader.SetUniform("uEquirectangularMap", 0);
+	glActiveTexture(GL_TEXTURE0);
+	HDRPanorama = (GLuint)(skyboxCmp->GetPanorama()->GetTextureProxy()->GetResourceID());
+	glBindTexture(GL_TEXTURE_2D, HDRPanorama);
+
+	glViewport(0, 0, 512, 512); // don't forget to configure the viewport to the capture dimensions.
```

Comment:
Ooops, looks like you forgot :)

Reply:
According to line 72 and 93 these are the capture dimensions. It may not change but we may move it to variable.